ktor-features/ktor-metrics-micrometer/jvm/src/io/ktor/metrics/micrometer/AbstractMicrometerMetrics.kt

```diff
@@ -0,0 +1,212 @@
+     * should enable these instead with [DistributionStatisticConfig.Builder.percentilesHistogram] as client side
+     * percentiles cannot be aggregated.
+     * @property timerBuilder can be used to configure each timer to add custom tags or configure individual SLAs etc
+     * */
+    open class AbstractConfiguration<TMeterRegistry : MeterRegistry> {
+
+        open lateinit var registry: TMeterRegistry
+
+        var meterBinders: List<MeterBinder> = listOf(
+            ClassLoaderMetrics(),
+            JvmMemoryMetrics(),
+            JvmGcMetrics(),
+            ProcessorMetrics(),
+            JvmThreadMetrics(),
+            FileDescriptorMetrics()
+        )
+
+        open var distributionStatisticConfig: DistributionStatisticConfig =
+            DistributionStatisticConfig.Builder()
+                .percentiles(0.5, 0.9, 0.95, 0.99)
+                .build()
+
+        internal var timerBuilder: Timer.Builder.(ApplicationCall, Throwable?) -> Unit = { _, _ -> }
+
+        fun timerBuilder(block: Timer.Builder.(ApplicationCall, Throwable?) -> Unit) {
+            timerBuilder = block
+        }
+
+        internal fun isRegistryInitialized() =
+            try {
+                registry
+                true
+            } catch (e: UninitializedPropertyAccessException) {
+                false
+            }
+
+    }
+
+
+    companion object {
+
+        private val measureKey = AttributeKey<CallMeasure>("metrics")
+
+        private const val baseName: String = "ktor.http.server"
+        const val requestTimerName = "$baseName.requests"
+        const val activeGaugeName = "$baseName.requests.active"
+    }
+
+    abstract class AbstractMetricsFeature<TMeterRegistry : MeterRegistry,
+        TConfiguration : AbstractConfiguration<TMeterRegistry>>(
+        private val constructor: (TConfiguration) -> AbstractMicrometerMetrics<TMeterRegistry, TConfiguration>,
+        private val configConstructor: () -> TConfiguration
+    ) :
+        ApplicationFeature<Application, TConfiguration, AbstractMicrometerMetrics<TMeterRegistry, TConfiguration>> {
+
+        override val key =
+            AttributeKey<AbstractMicrometerMetrics<TMeterRegistry, TConfiguration>>(
+                "metrics"
+            )
+
+        override fun install(pipeline: Application, configure: TConfiguration.() -> Unit):
+            AbstractMicrometerMetrics<TMeterRegistry, TConfiguration> {
+            val configuration = configConstructor().apply(configure)
+
+            if (!configuration.isRegistryInitialized()) {
+                throw IllegalArgumentException(
+                    "Meter registry is missing. Please initialize the field 'registry'"
+                )
+            }
+
+            val feature = constructor(configuration)
+
+            configuration.meterBinders.forEach { it.bindTo(configuration.registry) }
+
+            val phase = PipelinePhase("Metrics")
```

Comment:
Please note that only one metrics feature could be installed since double phase insertion is not allowed. Phase need to be renamed or installed only if not yet present.

Reply:
I prefixed it with the actual class name, this should make it unique someone needed more than one Metric feature.